Format URLs using a Jinja template macro

dkua opened the pull request:
This PR is a potential solution for the "clean URL" problem. By using a custom Jinja template macro to format any templated links we can ensure we have working links whether we're deployed on GitHub Pages or Cloudflare Pages.

Originally we thought that providing a base URL via the [`<base> tag`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) might be the solution to our "clean URL" problem. However after looking into it I realized that setting `<base>` is an all-or-nothing solution, meaning ALL non-absolute links will use the base URL even ones we didn't intend. For example `<base>` would break certain anchor links like `#`/`#top` for jumping to the top of the page without a workaround to allow `<base URL>#top` to be recognized as `#top`.

With the `fmturl` macro we have to be explicit about when we want a full URL to be rendered. The `fmturl(path)` macro will just return `path` if `base_url` is None or is not set at all. The Rust code will only set `base_url` to a non-None value when the `DEPLOY_BASE_URL` environment variable exists at runtime. This allows us to wrap all templated URLs with `fmturl` if we wish and only have the links change during the GitHub Pages deploy (via the Publish workflow). This PR only wraps the links inside the base template and the navbar, I didn't want to update every template file unless we were moving forward with this solution AND wanted to wrap every link.

## Rendering `base.html` without setting `base_url`
```bash
~/operation-mm [main] » minijinja-cli templates/shared/base.html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <title>Operation MM</title>
  <link href="public/images/squeaks-favicon.png" rel="icon">
  <link href="public/styles/reset.css" rel="stylesheet">

</head>

<body>

</body>

</html>
```

## Rendering `base.html` with `base_url` set to `github.com/`
```bash
~/operation-mm [main] » minijinja-cli templates/shared/base.html -D "base_url=github.com/"
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <title>Operation MM</title>
  <link href="github.com/public/images/squeaks-favicon.png" rel="icon">
  <link href="github.com/public/styles/reset.css" rel="stylesheet">

</head>

<body>

</body>

</html>
```

## Successful build using the `DEPLOY_BASE_URL` env var
https://github.com/dkua/operation-mm/actions/runs/12126803386/job/33809678961#step:5:10

This deploy is visible at https://dkua.github.io/operation-mm where you can see that the navbar links are fully rendered and the "clean URL" versions. The active page functionality on the navbar still works.

dkua commented:
@Trildar if you don't wish to enable GitHub Pages publishing from this repo you might want to disable Actions for now to avoid having these failing checks show up.

![image](https://github.com/user-attachments/assets/fdf78c35-c354-49e2-80e0-a77e56456f7b)


![image](https://github.com/user-attachments/assets/5b40e00b-cdb2-4924-a13f-bac45732e5c5)

Trildar commented:
Feel like it might be simpler to make the fragment links like `#top` include the page path (and implement `<base>`) rather than doing this. Could be done by passing in the `matched_path` like the code did previously.

Trildar commented on src/main.rs:
```diff
@@ -260,7 +256,8 @@ async fn messages(
     let env = template_engine.acquire_env()?;
-    let ctx = context! {messages};
+    let base_url = env::var("DEPLOY_BASE_URL").ok();
+    let ctx = context! { messages => messages, base_url => base_url };
```
Writing the key when it's the same as the variable name is redundant.

dkua replied:
Yes you're right, I'm not used to the field init shorthand and default to being explicit.

Trildar commented on templates/shared/header-nav.html:
```diff
@@ -1,14 +1,13 @@
-<a {% if page_current is startingwith page_path -%} class="active" {% endif -%}
-  href="{{path_link}}">{{display_text}}</a>
+{% macro link(page_path, display_text) %}
+<a {% if page_current==page_path -%} class="active" {% endif -%} href="{{ fmturl(page_path) }}">{{ display_text }}</a>
```
I kind of want to keep the current `page_current` matching logic.

dkua replied:
This isn't a blocking objection to the change as it doesn't affect the main functionality for this PR but isn't the current `page_current` matching logic redundant? What is the benefit of not matching `page_current` with the value used in `href` and having `page_link` and `page_path` be different?

Trildar replied:
It's admittedly semantics that don't matter for this project, but

1. Since the highlighting is now tied to the templates rather than the routes defined in axum, I felt like it made sense to just separate it completely.
2. The separation and matching on prefix theoretically allows for nested navigation styling, though that's not likely to be needed here.

Trildar commented on templates/shared/base.html:
```diff
@@ -1,11 +1,18 @@
+{% macro fmturl(path) -%}
+{% if base_url is defined and base_url -%}
```
`var is defined and var` is redundant. Can just do `if base_url`

dkua replied:
Yes you're right, I was only checking that the variable was defined initially and added `if base_url` after the None case came up when using `env::var().ok()` but didn't check if `is defined` was still needed.

dkua committed:
Add data

RikvanD1997 committed:
Home page changes

* Pat the Miku

* Pat image

* Fixed navbar, added YT video, styling fixes

---------

Co-authored-by: dkua <david@kua.io>

dkua committed:
Clean up Timeline related Javascript

Makes use of `DocumentFragment` to collect all the visible Timeline
Events before appending it to Timeline. This should reduce the number of
redraws that occur when filtering the Timeline.

Also change the `event-type` attribute to be `data-event-type`.

dkua committed:
Attempt a less jank grid for the markers

Instead of having a 0.5rem grid column with the markers on top we make
the column fit the markers and have the "line" drawn the same way its
done for the `baetag`.

By having the columns contain the markers we can make use of
`column-gap` to control the spacing between the markers and the other
boxes. Also lets us get rid of the jank transforms to centre the
markers.
